kani-driver/src/exe_trace.rs

```diff
@@ -0,0 +1,264 @@
+
+use crate::session::KaniSession;
+use kani_metadata::HarnessMetadata;
+use std::collections::hash_map::DefaultHasher;
+use std::ffi::OsString;
+use std::fs;
+use std::hash::{Hash, Hasher};
+use std::path::Path;
+use std::process::Command;
+
+/// Module for parsing deterministic values from CBMC output traces,
+/// generating executable traces, and adding them to the user's source code.
+
+impl KaniSession {
+    /// The main driver for generating executable traces and adding them to source code.
+    pub fn exe_trace_main(&self, specialized_obj: &Path, harness: &HarnessMetadata) {
+        if self.args.gen_exe_trace {
+            let (det_vals, interp_det_vals) = parser::get_det_vals(specialized_obj);
+            let (exe_trace, exe_trace_func_name) =
+                format_unit_test(&harness.mangled_name, &det_vals[..], &interp_det_vals[..]);
+
+            println!("Executable trace:\n");
+            println!("{}", exe_trace);
+
+            if self.args.add_exe_trace_to_src {
+                if !self.args.quiet {
+                    println!("Now modifying the source code to include the executable trace.");
+                }
+                let proof_harness_end_line: usize = harness
+                    .original_end_line
+                    .parse()
+                    .expect("Couldn't convert proof harness line from str to usize");
+                self.modify_src_code(
+                    &harness.original_file,
+                    proof_harness_end_line,
+                    &exe_trace,
+                    &exe_trace_func_name,
+                );
+            } else {
+                println!(
+                    "To automatically add this executable trace to the src code, run Kani with `--add-exe-trace-to-src`."
+                );
+            }
+        }
+    }
+
+    /// Add the exe trace to the user's source code, format it, and short circuit if code already present.
+    fn modify_src_code(
+        &self,
+        src_file_path: &str,
+        proof_harness_end_line: usize,
+        exe_trace: &str,
+        exe_trace_func_name: &str,
+    ) {
+        // Prep the source code and exec trace func.
+        let src_as_str =
+            fs::read_to_string(src_file_path).expect("Couldn't access proof harness source file");
+        let mut src_lines = src_as_str.split('\n').collect::<Vec<&str>>();
+        let mut exe_trace_lines = exe_trace.split('\n').collect::<Vec<&str>>();
+
+        // Short circuit if exe trace already in source code.
+        if src_as_str.contains(exe_trace_func_name) {
+            if !self.args.quiet {
+                println!("Exe trace already found in source code, so skipping modification.");
+            }
+            return;
+        }
+
+        // Calc inserted indexes and line numbers into source code to rustfmt only those lines.
+        // Indexes are into the vector (0-idx), lines are into src file (1-idx).
+        let start_line = proof_harness_end_line + 1;
```

Comment:
Why are you adding `+ 1` here and subtracting `1` after the fact?

Reply:
This code reasons about both indices and lines, as the comment alludes to. I found it more natural to first calculate the unit test start *line* (since I already have the harness end *line*). And then calculate the unit test start *idx*.

Reply:
Update: this is no longer here after I updated the logic in this function.